File: node/sample_09_cancel_order.js
/**
 * Sample 9 — Huy lenh
 * =====================
 * Dung phan khoi luong chua khop cua lenh dang mo.
 *
 * Luong:
 *   1. Lay so lenh, tim lenh dang PENDING/PARTIAL_FILLED
 *   2. Goi cancelOrderById kem accountNo/orderId
 *   3. API xac thuc quyen va trang thai lenh hien tai
 *   4. Neu hop le, he thong cap nhat CANCELLED cho phan chua khop
 *   5. Dong bo lai so lenh va so luong con treo
 */

import { Auth, Trading, OrderStatus } from '@ssi.developer/ssi-sdk';
import { config, ACCOUNT_NO, OTP } from './config.js';
import { ensureAuth } from './auth_helper.js';

const vnd = (x) => Number(x).toLocaleString('en-US', { maximumFractionDigits: 0 });

// Trang thai co the huy
const CANCELLABLE_STATUSES = new Set([
  OrderStatus.PENDING_APPROVAL,
  OrderStatus.READY,
  OrderStatus.SENT,
  OrderStatus.QUEUED,
  OrderStatus.PARTIAL_FILLED,
]);

const auth = new Auth(config);
await ensureAuth(auth, OTP);

const trading = new Trading(auth);

// --- Buoc 1: Lay so lenh, tim lenh dang mo ---
console.log('--- So lenh hom nay ---');
const orders = await trading.portfolio.getTodayOrders(ACCOUNT_NO);
const openOrders = orders.filter((o) => CANCELLABLE_STATUSES.has(o.status));
console.log(`Tong lenh: ${orders.length} | Lenh dang mo: ${openOrders.length}\n`);

if (openOrders.length === 0) {
  console.log('Khong co lenh nao dang mo de huy.');

  // --- Response Summary ---
  console.log('\n[Response] open_count|cancel_status');
  console.log('0|N/A');
} else {
  for (const order of openOrders) {
    const remaining = order.quantity - order.filledQuantity;
    console.log(
      `  OrderID: ${order.orderId} | ${order.symbol} ${order.side} ${order.orderType} | ` +
      `SL: ${order.quantity} @ ${order.price} | Khop: ${order.filledQuantity} | ` +
      `Con: ${remaining} | Status: ${order.status}`,
    );
  }

  // --- Buoc 2: Huy lenh dau tien trong danh sach ---
  const target = openOrders[0];
  console.log(`\n--- Huy lenh: ${target.orderId} ---`);
  const result = await trading.trading.cancelOrderById(ACCOUNT_NO, target.orderId);
  console.log('  Ket qua huy:', result);

  // --- Buoc 3: Xac nhan trang thai sau huy ---
  console.log('\n--- Kiem tra so lenh sau huy ---');
  const ordersAfter = await trading.portfolio.getTodayOrders(ACCOUNT_NO);
  for (const order of ordersAfter) {
    if (order.orderId === target.orderId) {
      console.log(
        `  OrderID: ${order.orderId} | Status: ${order.status} | ` +
        `Khop: ${order.filledQuantity} | Da huy: ${order.cancelQuantity}`,
      );
      break;
    }
  }

  // --- Response Summary ---
  console.log('\n[Response] open_count|cancel_status');
  console.log(`${openOrders.length}|${result.status}`);
}

// --- Buoc 4: Cap nhat lai so du ---
console.log('\n--- So du sau huy ---');
const balance = await trading.portfolio.getEquityBalance(ACCOUNT_NO);
console.log(`  Tien mat kha dung: ${vnd(balance.accountBalance).padStart(15)}`);

File: dotnet/Sample09CancelOrder.cs
// Sample 09 — Hủy lệnh
// Dung phan khoi luong chua khop cua lenh dang mo.

using SsiSdk;

public static class Sample09CancelOrder
{
    private static readonly HashSet<string> CancellableStatuses = new()
    {
        OrderStatus.PendingApproval,
        OrderStatus.Ready,
        OrderStatus.Sent,
        OrderStatus.Queued,
        OrderStatus.PartialFilled,
    };

    public static async Task RunAsync()
    {
        var config = SampleConfig.Create();
        using var auth = new AuthClient(config);
        await AuthHelper.EnsureAuthAsync(auth, SampleConfig.Otp);

        var trading = new TradingClient(auth);

        // --- Buoc 1: Lay so lenh, tim lenh dang mo ---
        Console.WriteLine("--- So lenh hom nay ---");
        var orders = await trading.Portfolio.GetTodayOrdersAsync(SampleConfig.AccountNo);

        var openOrders = orders.Where(o => CancellableStatuses.Contains(o.Status)).ToList();
        Console.WriteLine($"Tong lenh: {orders.Count} | Lenh dang mo: {openOrders.Count}\n");

        if (openOrders.Count == 0)
        {
            Console.WriteLine("Khong co lenh nao dang mo de huy.");

            // --- Response Summary ---
            Console.WriteLine("\n[Response] open_count|cancel_status");
            Console.WriteLine("0|N/A");
            return;
        }

        foreach (var order in openOrders)
        {
            var remaining = order.Quantity - order.FilledQuantity;
            Console.WriteLine(
                $"  OrderID: {order.OrderId} | {order.Symbol} {order.Side} " +
                $"{order.OrderType} | SL: {order.Quantity} @ {order.Price} " +
                $"| Khop: {order.FilledQuantity} | Con: {remaining} " +
                $"| Status: {order.Status}");
        }

        // --- Buoc 2: Huy lenh dau tien trong danh sach ---
        var target = openOrders[0];
        Console.WriteLine($"\n--- Huy lenh: {target.OrderId} ---");

        var result = await trading.Trading.CancelOrderByOrderIdAsync(
            SampleConfig.AccountNo, target.OrderId);
        Console.WriteLine($"  Ket qua huy: OrderId={result.OrderId} Status={result.Status}");

        // --- Buoc 3: Xac nhan trang thai sau huy ---
        Console.WriteLine("\n--- Kiem tra so lenh sau huy ---");
        var ordersAfter = await trading.Portfolio.GetTodayOrdersAsync(SampleConfig.AccountNo);
        foreach (var order in ordersAfter)
        {
            if (order.OrderId == target.OrderId)
            {
                Console.WriteLine(
                    $"  OrderID: {order.OrderId} | " +
                    $"Status: {order.Status} | " +
                    $"Khop: {order.FilledQuantity} | " +
                    $"Da huy: {order.CancelQuantity}");
                break;
            }
        }

        // --- Response Summary ---
        Console.WriteLine("\n[Response] open_count|cancel_status");
        Console.WriteLine($"{openOrders.Count}|{result.Status}");

        // --- Buoc 4: Cap nhat lai so du ---
        Console.WriteLine("\n--- So du sau huy ---");
        var balance = await trading.Portfolio.GetEquityBalanceAsync(SampleConfig.AccountNo);
        if (balance is not null)
        {
            Console.WriteLine($"  Tien mat kha dung: {balance.AccountBalance,15:N0}");
        }
    }
}
